tket/src/Placement/NeighbourPlacements.cpp

```diff
@@ -0,0 +1,105 @@
+#include "NeighbourPlacements.hpp"
+
+#include <cstdlib>
+
+#include "TokenSwapping/SwapListOptimiser.hpp"
+
+namespace tket {
+
+NeighbourPlacements::NeighbourPlacements(
+    const Architecture& arc, const qubit_mapping_t& init_map)
+    : arc_(arc), init_map_(init_map), u_to_node_(), rng_() {
+  auto nodes = arc_.get_all_nodes_vec();
+  for (unsigned i = 0; i < nodes.size(); ++i) {
+    u_to_node_.left.insert({i, nodes[i]});
+  }
+}
+
+NeighbourPlacements::ResultVec NeighbourPlacements::get(
+    unsigned dist, unsigned n, bool optimise, unsigned seed) {
+  rng_.set_seed(seed);
+  ResultVec res;
+  for (unsigned i = 0; i < n; ++i) {
+    res.push_back(gen_result(dist, optimise));
+  }
+  return res;
+}
+
+NeighbourPlacements::Result NeighbourPlacements::gen_result(
+    unsigned dist, bool optimise) {
+  SwapList swaps;
+  tsa_internal::SwapListOptimiser optimiser;
+
+  // it might be impossible to find `dist` non-trivial swaps
+  unsigned n_unsuccessful = 0;
+  const unsigned N_TRIES = 10;
+
+  while (swaps.size() < dist && n_unsuccessful < N_TRIES) {
+    unsigned d = dist - swaps.size();
+    SwapList new_swaps = gen_swap_list(d);
+
+    if (optimise) {
+      SwapList swaps_candidate = swaps;
+      for (auto swap : new_swaps.to_vector()) {
+        swaps_candidate.push_back(swap);
+      }
+      optimiser.full_optimise(swaps_candidate);
+      if (swaps_candidate.size() > swaps.size()) {
+        swaps = std::move(swaps_candidate);
+        n_unsuccessful = 0;
+      } else {
+        ++n_unsuccessful;
+      }
+    } else {
+      for (auto swap : new_swaps.to_vector()) {
+        swaps.push_back(swap);
+      }
+    }
+  }
+
+  if (n_unsuccessful == N_TRIES) {
+    throw NotValid(
+        "Unable to generate " + std::to_string(dist) +
+        " swaps for given architecture");
+  }
+
+  return convert_to_res(swaps.to_vector());
+}
+
+SwapList NeighbourPlacements::gen_swap_list(unsigned dist) {
+  SwapList swaps;
+  auto edges = arc_.get_all_edges_vec();
+  unsigned m = edges.size();
+  for (unsigned i = 0; i < dist; ++i) {
+    auto [n1, n2] = edges[rng_.get_size_t(m - 1)];
+    Swap new_swap{u_to_node_.right.at(n1), u_to_node_.right.at(n2)};
```

Comment:
It looks like to me that the second swap could be identical to the first one? have you thought about avoiding this case?

Reply:
Yes. There are lots of redundancies that are possible. I take care of them by using Zen's optimiser that simplifies a chain of swaps (and will definitely cancel two swaps that are identical). Whatever I would do here, it won't be better than what Zen has already implemented so I decided to leave all simplification with his code.

Reply:
Sound valid, lets keep it as it is.